rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -600,68 +659,52 @@ be reused in a TF op (as the downstream of the Dataset, or in a Python wrapper).
-    the TF trainer. As that is a direct result of I/O + parsing being abstracted
-    out.
-
-The disadvantages of this approach are:
-
-*   This tightly couples TF with TFX components, in the following ways:
-
-    *   Components will need TF to read the data in.
-    *   Components that analyze the data (e.g. TFDV) will operate against
-        StructuredTensors, and the easiest way to to conduct certain
-        computations (for example, slicing a StructuredTensor or computing the
-        mean) with StructuredTensors is through TF ops and their python
-        bindings.
-
-    Such a coupling is not in our favor because:
-
-    *   Some TFX components functionally do not require TF to work. For example,
-        TFDV can analyze any data set. TFMA can analyze a model while treating
-        the model as a blackbox. In both cases, the ML library that trained the
-        model is irrelevant and TF should not be assumed. And coupling with TF
-        does not only introduce a heavy dependency, but also forces the user to
-        learn about TF if they need to implement a TFXIO for their data format.
-
-    *   Operations against StructuredTensors boils down to Python + TF ops. The
-        overhead of either is much higher than just calling an Arrow C++ API
-        that does the same operation.
-
-    *   The main extension point of TF is Ops which don't understand the nested
-        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
-        deals with StructuredTensors will be much complicated.
+## StructuredTensor
+
+We’ve considered an alternative where [StructuredTensor](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, but it does not meet all of the
+requirements:
+
+|                                   | Arrow RecordBatch  | StructuredTensor  |
+| :-------------------------------- | :----------------- | :---------------- |
+| Columnar                          | Yes                | Yes               |
+| Lossless encoding (nullity)       | Yes                | No (see remark 1) |
+| Efficient translatable to Tensors | Yes (see remark 2) | Yes               |
+| Efficient slicing, filtering and concatenation | Yes                | No (see remark 3) |
+| Interoperability with the rest of the world | Good through Apache Arrow | Needs adaptation (see remark 4) |
+
+Remarks:
+
+1.  We could revise the design of StructuredTensor to include the nullibility
+    support.
+2.  Only when the backing buffers are aligned correctly. Currently both TF
+    and Apache Arrow has 64-byte alignment. And this can be enforced by
+    implementing our own Arrow MemoryPool wrapping a TF allocator.
+3.  See the comparison in [this colab notebook](https://colab.sandbox.google.com/drive/1SCQs88J4Tc6HKk2AfpYvELaJDxa4h4IQ).
+    *   It’s worth calling out that Arrow is meant to be a data analysis library
```

Comment:
I am not the expert here, but @jsimsa @edloper perhaps can comment on the particular comparisons made in the notebook.

Reply:
Also @mzinkevi

Reply:
I believe adding slicing, filtering, etc into StructuredTensor is very easy.

Reply:
let's take RaggedTensor as an example -- I would assume StructuredTensor's implementation of those operations would not be faster than RaggedTensors (as it would rely on RaggedTensor's). Indeed, there is already built-in slicing and concatenating support, but tf.concat on RaggedTensors is 100x (or 200x in Eager mode) slower than pa.concat_arrays() on Arrow ListArrays [1].

[1] https://colab.sandbox.google.com/drive/1wGfmNnkVDFtq48d1RTSaEkxePPBJ0go4#scrollTo=U7eareQUdKWN

Reply:
That sounds like a performance but on raggedtensor which should be fixable.